File: KannadaNudiAndroid/app/src/main/java/com/kannadanudi/keyboard/TransliterationEngine.kt
package com.kannadanudi.keyboard

import java.lang.StringBuilder

enum class KeyboardLayout {
    Nudi,
    Baraha,
    English
}

data class TransliterationResult(val text: String, val backspaceCount: Int)

class TransliterationEngine {
    private val buffer = StringBuilder()
    var currentLayout = KeyboardLayout.Baraha

    // Nudi Map (Direct Layout) - Independent Chars
    private val nudiMap = mapOf(
        // Top Row
        "q" to "ಟ", "Q" to "ಠ",
        "w" to "ಡ", "W" to "ಢ",
        "e" to "ಎ", "E" to "ಏ",
        "r" to "ರ", "R" to "ಋ",
        "t" to "ತ", "T" to "ಥ",
        "y" to "ಯ", "Y" to "ಐ",
        "u" to "ಉ", "U" to "ಊ",
        "i" to "ಇ", "I" to "ಈ",
        "o" to "ಒ", "O" to "ಓ",
        "p" to "ಪ", "P" to "ಫ",

        // Middle Row
        "a" to "ಅ", "A" to "ಆ",
        "s" to "ಸ", "S" to "ಶ",
        "d" to "ದ", "D" to "ಧ",
        "f" to "್", "F" to "್",
        "g" to "ಗ", "G" to "ಘ",
        "h" to "ಹ", "H" to "ಃ", // Visarga
        "j" to "ಜ್", "J" to "ಝ್",
        // User list J=ಝ (Direct). Wait, my previous write said "j" to "ಜ್" (Halant)?
        // I will fix it to "ಜ" and "ಝ" (Direct) as per list.
        "j" to "ಜ", "J" to "ಝ",
        "k" to "ಕ", "K" to "ಖ",
        "l" to "ಲ", "L" to "ಳ",

        // Bottom Row
        "z" to "ಞ", "Z" to "ಙ",
        "x" to "ಷ", "X" to "ಷ",
        "c" to "ಚ", "C" to "ಚ",
        "v" to "ವ", "V" to "ವ",
        "b" to "ಬ", "B" to "ಬ",
        "n" to "ನ", "N" to "ಣ",
        "m" to "ಮ", "M" to "ಮ"
    )

    // Matra Map (Vowel Signs)
    private val nudiMatras = mapOf(
        "A" to "ಾ",
        "i" to "ಿ", "I" to "ೀ",
        "u" to "ು", "U" to "ೂ",
        "R" to "ೃ",
        "e" to "ೆ", "E" to "ೇ",
        "Y" to "ೈ", // Shift+y = I -> Matra ai
        "o" to "ೊ", "O" to "ೋ"
        // 'a' has no matra (implicit)
    )

    // Baraha Map (Phonetic)
    private val barahaMap = mapOf(
        // Consonants (Halant default)
        "k" to "ಕ್", "K" to "ಖ್", "g" to "ಗ್", "G" to "ಘ್", "ng" to "ಂಗ್",
        "c" to "ಚ್", "ch" to "ಚ್", "C" to "ಛ್", "Ch" to "ಛ್", "j" to "ಜ್", "J" to "ಝ್", "nj" to "ಞ್",
        "T" to "ಟ್", "Th" to "ಠ್", "D" to "ಡ್", "Dh" to "ಢ್", "N" to "ಣ್",
        "t" to "ತ್", "th" to "ಥ್", "d" to "ದ್", "dh" to "ಧ್", "n" to "ನ್",
        "p" to "ಪ್", "P" to "ಫ್", "f" to "ಫ್", "b" to "ಬ್", "B" to "ಭ್", "m" to "ಮ್",
        "y" to "ಯ್", "r" to "ರ್", "l" to "ಲ್", "v" to "ವ್", "w" to "ವ್",
        "S" to "ಶ್", "sh" to "ಷ್", "s" to "ಸ್", "h" to "ಹ್",
        "L" to "ಳ್",

        // Vowels (Independent)
        "a" to "ಅ", "aa" to "ಆ", "A" to "ಆ",
        "i" to "ಇ", "ii" to "ಈ", "I" to "ಈ",
        "u" to "ಉ", "uu" to "ಊ", "U" to "ಊ",
        "R" to "ಋ", "Ru" to "ಋ",
        "e" to "ಎ", "ee" to "ಏ", "E" to "ಏ",
        "ai" to "ಐ",
        "o" to "ಒ", "oo" to "ಓ", "O" to "ಓ",
        "au" to "ಔ", "ou" to "ಔ",

        // Modifiers
        "M" to "ಂ", "H" to "ಃ"
    )

    private val barahaVowelSigns = mapOf(
        "a" to "",
        "aa" to "ಾ", "A" to "ಾ",
        "i" to "ಿ",
        "ii" to "ೀ", "I" to "ೀ",
        "u" to "ು",
        "uu" to "ೂ", "U" to "ೂ",
        "R" to "ೃ", "Ru" to "ೃ",
        "e" to "ೆ",
        "ee" to "ೇ", "E" to "ೇ",
        "ai" to "ೈ",
        "o" to "ೊ",
        "oo" to "ೋ", "O" to "ೋ",
        "au" to "ೌ", "ou" to "ೌ"
    )

    fun setLayout(layout: KeyboardLayout) {
        currentLayout = layout
        clearBuffer()
    }

    fun clearBuffer() {
        buffer.setLength(0)
    }

    fun removeLast() {
        if (buffer.isNotEmpty()) {
            buffer.setLength(buffer.length - 1)
        }
    }

    fun getTransliteration(key: String, lastCommittedChar: Char? = null): TransliterationResult {
        if (key.isEmpty()) return TransliterationResult("", 0)

        return when (currentLayout) {
            KeyboardLayout.Nudi -> getNudiTransliteration(key, lastCommittedChar)
            KeyboardLayout.Baraha -> getBarahaTransliteration(key)
            KeyboardLayout.English -> {
                buffer.setLength(0)
                TransliterationResult(key, 0)
            }
        }
    }

    private fun getNudiTransliteration(key: String, lastCommittedChar: Char?): TransliterationResult {
        // Direct Mapping with Matra Composition Context

        // 1. Check if key is a vowel that should become a Matra
        if (lastCommittedChar != null && isKannadaConsonant(lastCommittedChar) && nudiMatras.containsKey(key)) {
            // Return Matra
            return TransliterationResult(nudiMatras[key]!!, 0)
        }

        // 2. Default: Map to Independent Char
        if (nudiMap.containsKey(key)) {
            buffer.setLength(0)
            return TransliterationResult(nudiMap[key]!!, 0)
        }

        // Pass through if not found
        buffer.setLength(0)
        return TransliterationResult(key, 0)
    }

    private fun isKannadaConsonant(c: Char): Boolean {
        // Range for Kannada Consonants: 0x0C95 (ka) to 0x0CB9 (ha)
        // This covers most consonants.
        // Also includes 0x0CB3 (Lla)
        // Excludes 0x0C82..0x0C94 (Vowels, signs)
        // Excludes 0x0CBE.. (Matras)

        val code = c.code
        return (code in 0x0C95..0x0CB9)
    }

    private fun getBarahaTransliteration(key: String): TransliterationResult {
        val combinedKey = buffer.toString() + key

        if (buffer.isNotEmpty()) {
            // 1. Try to match longest sequence backwards for VOWEL MODIFIERS on CONSONANTS
            for (i in combinedKey.length - 1 downTo 0) {
                val potentialConsonantToken = combinedKey.substring(0, i)
                val potentialVowelToken = combinedKey.substring(i)

                if (barahaMap.containsKey(potentialConsonantToken) && isBarahaConsonant(potentialConsonantToken)) {
                    if (barahaVowelSigns.containsKey(potentialVowelToken)) {
                        // Found C+V combo
                        val previousOutput = recalculateOutput(buffer.toString())

                        val consChar = barahaMap[potentialConsonantToken]!!
                        val baseChar = consChar.trimEnd('\u0CCD')
                        val matra = barahaVowelSigns[potentialVowelToken]!!

                        val replacement = baseChar + matra

                        buffer.append(key)
                        return TransliterationResult(replacement, previousOutput.length)
                    }
                }
            }

            // 2. Check if combined key is a valid Consonant or Vowel (Extension)
            if (barahaMap.containsKey(combinedKey)) {
                val previousOutput = recalculateOutput(buffer.toString())
                buffer.setLength(0)
                buffer.append(combinedKey)
                return TransliterationResult(barahaMap[combinedKey]!!, previousOutput.length)
            }
        }

        // Standard processing
        if (barahaMap.containsKey(key)) {
            buffer.setLength(0)
            buffer.append(key)
            return TransliterationResult(barahaMap[key]!!, 0)
        }

        // Unmapped
        buffer.setLength(0)
        buffer.append(key)
        return TransliterationResult(key, 0)
    }

    private fun recalculateOutput(bufferKeys: String): String {
        if (bufferKeys.isEmpty()) return ""

        // Try to match C+V
        for (i in bufferKeys.length - 1 downTo 0) {
            val c = bufferKeys.substring(0, i)
            val v = bufferKeys.substring(i)

            if (barahaMap.containsKey(c) && isBarahaConsonant(c) && barahaVowelSigns.containsKey(v)) {
                val cons = barahaMap[c]!!
                val baseChar = cons.trimEnd('\u0CCD')
                val matra = barahaVowelSigns[v]!!
                return baseChar + matra
            }
        }

        // Try match full key
        if (barahaMap.containsKey(bufferKeys)) {
            return barahaMap[bufferKeys]!!
        }

        return ""
    }

    private fun isBarahaConsonant(k: String): Boolean {
        val value = barahaMap[k] ?: return false
        return value.endsWith("\u0CCD")
    }
}

File: KannadaNudi-iOS/KannadaKeyboard/Logic/TransliterationEngine.swift
import Foundation

enum KeyboardLayout {
    case nudi
    case baraha
}

struct TransliterationResult {
    let text: String
    let backspaceCount: Int
}

class TransliterationEngine {
    private var buffer = ""
    var currentLayout: KeyboardLayout = .baraha

    // MARK: - Maps

    // Nudi Map (Direct/Legacy)
    private let nudiMap: [String: String] = [
        // Consonants
        "k": "ಕ್", "g": "ಗ್", "c": "ಚ್", "j": "ಜ್",
        "t": "ಟ್", "d": "ಡ್", "N": "ಣ್",
        "w": "ತ್", "q": "ದ್", "n": "ನ್",
        "p": "ಪ್", "b": "ಬ್", "m": "ಮ್",
        "y": "ಯ್", "r": "ರ್", "l": "ಲ್", "v": "ವ್",
        "s": "ಸ್", "h": "ಹ್", "L": "ಳ್",

        // Vowels
        "a": "ಅ", "A": "ಆ", "i": "ಇ", "I": "ಈ",
        "u": "ಉ", "U": "ಊ", "e": "ಎ", "E": "ಏ",
        "o": "ಒ", "O": "ಓ"
    ]

    private let nudiVowelSigns: [String: String] = [
        "a": "", // 'a' removes halant
        "A": "ಾ",
        "i": "ಿ",
        "I": "ೀ",
        "u": "ು",
        "U": "ೂ",
        "e": "ೆ",
        "E": "ೇ",
        "o": "ೊ",
        "O": "ೋ"
    ]

    // Baraha Map (Phonetic)
    private let barahaMap: [String: String] = [
        // Consonants (Halant default)
        "k": "ಕ್", "K": "ಖ್", "g": "ಗ್", "G": "ಘ್", "ng": "ಂಗ್",
        "c": "ಚ್", "ch": "ಚ್", "C": "ಛ್", "Ch": "ಛ್", "j": "ಜ್", "J": "ಝ್", "nj": "ಞ್",
        "T": "ಟ್", "Th": "ಠ್", "D": "ಡ್", "Dh": "ಢ್", "N": "ಣ್",
        "t": "ತ್", "th": "ಥ್", "d": "ದ್", "dh": "ಧ್", "n": "ನ್",
        "p": "ಪ್", "P": "ಫ್", "f": "ಫ್", "b": "ಬ್", "B": "ಭ್", "m": "ಮ್",
        "y": "ಯ್", "r": "ರ್", "l": "ಲ್", "v": "ವ್", "w": "ವ್",
        "S": "ಶ್", "sh": "ಷ್", "s": "ಸ್", "h": "ಹ್",
        "L": "ಳ್",

        // Vowels (Independent)
        "a": "ಅ", "aa": "ಆ", "A": "ಆ",
        "i": "ಇ", "ii": "ಈ", "I": "ಈ",
        "u": "ಉ", "uu": "ಊ", "U": "ಊ",
        "R": "ಋ", "Ru": "ಋ",
        "e": "ಎ", "ee": "ಏ", "E": "ಏ",
        "ai": "ಐ",
        "o": "ಒ", "oo": "ಓ", "O": "ಓ",
        "au": "ಔ", "ou": "ಔ",

        // Modifiers
        "M": "ಂ", "H": "ಃ"
    ]

    private let barahaVowelSigns: [String: String] = [
        "a": "",
        "aa": "ಾ", "A": "ಾ",
        "i": "ಿ",
        "ii": "ೀ", "I": "ೀ",
        "u": "ು",
        "uu": "ೂ", "U": "ೂ",
        "R": "ೃ", "Ru": "ೃ",
        "e": "ೆ",
        "ee": "ೇ", "E": "ೇ",
        "ai": "ೈ",
        "o": "ೊ",
        "oo": "ೋ", "O": "ೋ",
        "au": "ೌ", "ou": "ೌ"
    ]

    // MARK: - Methods

    func setLayout(_ layout: KeyboardLayout) {
        currentLayout = layout
        clearBuffer()
    }

    func clearBuffer() {
        buffer = ""
    }

    func removeLast() {
        if !buffer.isEmpty {
            buffer.removeLast()
        }
    }

    func getTransliteration(key: String) -> TransliterationResult {
        if key.isEmpty {
            return TransliterationResult(text: "", backspaceCount: 0)
        }

        if currentLayout == .nudi {
            return getNudiTransliteration(key: key)
        } else {
            return getBarahaTransliteration(key: key)
        }
    }

    private func getNudiTransliteration(key: String) -> TransliterationResult {
        // Nudi/KGP Legacy Logic:

        // If key is a vowel modifier and buffer ends in consonant key
        if nudiVowelSigns.keys.contains(key), !buffer.isEmpty {
            let lastKey = String(buffer.last!)
            if isNudiConsonantKey(lastKey) {
                if let halantForm = nudiMap[lastKey], let sign = nudiVowelSigns[key] {
                    // halantForm e.g. "ಕ್" (0C95 0CCD)
                    let baseForm = halantForm.trimmingCharacters(in: CharacterSet(charactersIn: "\u{0CCD}"))
                    let replacement = baseForm + sign

                    // Swift string count might count grapheme clusters, but input proxy usually deletes by character code units?
                    // Actually, iOS backspace count is usually by character.
                    // "ಕ್" is 2 chars.
                    let removeCount = halantForm.count

                    buffer.append(key)
                    return TransliterationResult(text: replacement, backspaceCount: removeCount)
                }
            }
        }

        // If key is a consonant
        if let val = nudiMap[key] {
            buffer.append(key)
            return TransliterationResult(text: val, backspaceCount: 0)
        }

        // Default
        buffer = ""
        buffer.append(key)
        return TransliterationResult(text: key, backspaceCount: 0)
    }

    private func getBarahaTransliteration(key: String) -> TransliterationResult {
        let combinedKey = buffer + key

        if !buffer.isEmpty {
            // 1. Try to match longest sequence backwards for VOWEL MODIFIERS on CONSONANTS
            let combinedCount = combinedKey.count
            // Swift strings are tricky with indices. Using integer loops for easier porting logic.
            // But we need to handle unicode properly if keys are unicode. Assuming keys are ASCII/single chars for now as per maps.

            for i in (0..<combinedCount).reversed() {
                // i is the split point.
                // substring(0, i) -> prefix
                // substring(i) -> suffix

                let index = combinedKey.index(combinedKey.startIndex, offsetBy: i)
                let potentialConsonantToken = String(combinedKey[..<index])
                let potentialVowelToken = String(combinedKey[index...])

                if let _ = barahaMap[potentialConsonantToken], isBarahaConsonant(potentialConsonantToken) {
                    if let matra = barahaVowelSigns[potentialVowelToken] {
                        // Found C+V combo
                        let previousOutput = recalculateOutput(bufferKeys: buffer)

                        let consChar = barahaMap[potentialConsonantToken]!
                        let baseChar = consChar.trimmingCharacters(in: CharacterSet(charactersIn: "\u{0CCD}"))

                        let replacement = baseChar + matra

                        buffer.append(key)
                        return TransliterationResult(text: replacement, backspaceCount: previousOutput.count)
                    }
                }
            }

            // 2. Check if combined key is a valid Consonant or Vowel (Extension)
            if let val = barahaMap[combinedKey] {
                let previousOutput = recalculateOutput(bufferKeys: buffer)
                buffer = ""
                buffer.append(combinedKey)
                return TransliterationResult(text: val, backspaceCount: previousOutput.count)
            }
        }

        // Standard processing
        if let val = barahaMap[key] {
            buffer = ""
            buffer.append(key)
            return TransliterationResult(text: val, backspaceCount: 0)
        }

        // Unmapped
        buffer = ""
        buffer.append(key)
        return TransliterationResult(text: key, backspaceCount: 0)
    }

    private func recalculateOutput(bufferKeys: String) -> String {
        if bufferKeys.isEmpty { return "" }

        let count = bufferKeys.count
        for i in (0..<count).reversed() {
            let index = bufferKeys.index(bufferKeys.startIndex, offsetBy: i)
            let c = String(bufferKeys[..<index])
            let v = String(bufferKeys[index...])

            if let _ = barahaMap[c], isBarahaConsonant(c), let matra = barahaVowelSigns[v] {
                let cons = barahaMap[c]!
                let baseChar = cons.trimmingCharacters(in: CharacterSet(charactersIn: "\u{0CCD}"))
                return baseChar + matra
            }
        }

        if let val = barahaMap[bufferKeys] {
            return val
        }

        return ""
    }

    private func isNudiConsonantKey(_ k: String) -> Bool {
        return nudiMap.keys.contains(k) && !nudiVowelSigns.keys.contains(k)
    }

    private func isBarahaConsonant(_ k: String) -> Bool {
        guard let value = barahaMap[k] else { return false }
        return value.hasSuffix("\u{0CCD}")
    }
}
